src/GameLogic/Level.ts

```diff
@@ -33,4 +34,19 @@ export class Level {
 		this.tilesFloor = new Grid2D<FloorType>(this.width, this.height, FloorType.FloorTile);
 		this.entities = new Entities();
 	}
+
+	public clone(): Level {
+		const clonedLevel = new Level({
+			width: this.width,
+			height: this.height,
+			playerStartX: this.playerStartX,
+			playerStartY: this.playerStartY,
+		});
+		clonedLevel.tilesFloor.setAllByCallback((x, y) => this.tilesFloor.get(x, y));
+		const clonedEntities: Entity[] = [];
+		this.entities.entities.forEach(entity => clonedEntities.push(entity.clone()));
```

Comment:
Instead of creating the array and then replacing `Entities` (which should be a readonly field to avoid doing this) why not just have `clonedLevel.entities.addEntity(entity.clone())` in the `forEach`?

And the reason why I'd rather make that variable readonly: replacing the whole `Entities` object is not really required as part of the API, as there aren't any real benefits for having it that way but there are ways to accidentally trip on it - for example by assinging another level's entities to it and then wonder why things change when they shouldn't.